vibin/mediaservers/cxnv2.py

```diff
@@ -0,0 +1,589 @@
+
+    def create_track(self, result: _BrowseResult) -> Track:
+        """Creates a new Track with information from this _BrowseResult"""
+        return Track(
+            id = self.id_map.get_id(result.path),
+            albumId = self.id_map.get_id(result.parent_path),
+            title = result.title,
+            artist = result.artist,
+            album = result.album or "Unknown Album",
+            duration = result.duration,
+            genre = result.genre,
+            album_art_uri = self.artwork.get(result.album_art_uri)
+                            or result.album_art_uri,
+            original_track_number = result.original_track_number,
+        )
+
+    def create_album(self, result: _BrowseResult) -> Album:
+        """Creates a new Album with information from this _BrowseResult."""
+        album_id = self.id_map.get_id(result.path)
+        return Album(
+            id = album_id,
+            title = result.title,
+            artist = result.artist or self._guess_artist(album_id),
+            genre = result.genre or self._guess_genre(album_id),
+            album_art_uri = self.artwork.get(result.album_art_uri)
+                            or result.album_art_uri,
+        )
+
+    def _guess_artist(self, album_id: MediaId):
+        """Infers the artist name from an album's tracks.
+
+        If a single artist is responsible for most of the tracks, assume the
+        album is credited to that artist. Otherwise, use the name `Various
+        Artists`.
+        """
+        tracks = self.tracks_by_album_id[album_id]
+        if not tracks:
+            return "Unknown"
+        (artist, count) = Counter([t.artist for t in tracks]).most_common(1)[0]
+        if artist and count > len(tracks) / 2:
+            return artist
+        else:
+            return "Various Artists"
+
+    def _guess_genre(self, album_id: MediaId):
+        """Infers a genre from an album's tracks.
+
+        If most of the tracks share a genre, uses that genre. Otherwise, return
+        the string 'Unknown' which the UI expects in the absense of a genre.
+        """
+        tracks = self.tracks_by_album_id[album_id]
+        if not tracks:
+            return "Unknown"
+        (genre, count) = Counter([t.genre for t in tracks]).most_common(1)[0]
+        if genre and count > len(tracks) / 2:
+            return genre
+        else:
+            return "Unknown"
+
+    def create_media_folder(self, result: _BrowseResult) -> MediaFolder:
+        """Creates a new MediaFolder with information from this _BrowseResult."""
+        return MediaFolder(
+            title = result.title,
+            album_art_uri = result.album_art_uri,
+            artist = result.artist,
+            genre = result.genre,
+        )
+
+    def create_model_object(self, result: _BrowseResult) -> Track | Album | MediaFolder:
+        """Returns a model object for this _BrowseResult."""
+        match result.type:
```

Comment:
Love seeing the `match` syntax here. I need to internalize this language feature.

Reply:
Another feature that didn't exist last time I used Python :)